File: examples/gray-scott.jl
# found in examples/gray-scott.jl
# using cuNumeric
using Plots

struct Params
    dx::Float64
    dt::Float64
    c_u::Float64
    c_v::Float64
    f::Float64
    k::Float64

    function Params(dx=1, c_u=1.0, c_v=0.3, f=0.03, k=0.06)
        new(dx, 1e-4, c_u, c_v, f, k)
    end
end

function step(u, v, u_new, v_new, args::Params)
    # calculate F_u and F_v functions
    # currently we don't have NDArray^x working yet. 
    F_u = (-u[2:end-1, 2:end-1]*(v[2:end-1, 2:end-1] * v[2:end-1, 2:end-1])) + args.f*(1 .- u[2:end-1, 2:end-1])
    F_v = (u[2:end-1, 2:end-1]*(v[2:end-1, 2:end-1] * v[2:end-1, 2:end-1])) - (args.f+args.k)*v[2:end-1, 2:end-1]
    
    # 2-D Laplacian of f using array slicing, excluding boundaries
    # For an N x N array f, f_lap is the Nend x Nend array in the "middle"
    u_lap = ((u[3:end, 2:end-1] - 2*u[2:end-1, 2:end-1] + u[1:end-2, 2:end-1]) ./ args.dx^2 
           + (u[2:end-1, 3:end] - 2*u[2:end-1, 2:end-1] + u[2:end-1, 1:end-2]) ./ args.dx^2)
    v_lap = ((v[3:end, 2:end-1] - 2*v[2:end-1, 2:end-1] + v[1:end-2, 2:end-1]) ./ args.dx^2 
           + (v[2:end-1, 3:end] - 2*v[2:end-1, 2:end-1] + v[2:end-1, 1:end-2]) ./ args.dx^2)

    # Forward-Euler time step for all points except the boundaries
    u_new[2:end-1, 2:end-1] = ((args.c_u * u_lap) + F_u) * args.dt + u[2:end-1, 2:end-1]
    v_new[2:end-1, 2:end-1] = ((args.c_v * v_lap) + F_v) * args.dt + v[2:end-1, 2:end-1]

    # Apply periodic boundary conditions
    u_new[:,1] = u[:,end-1]
    u_new[:,end] = u[:,2]
    u_new[1,:] = u[end-1,:]
    u_new[end,:] = u[2,:]
    v_new[:,1] = v[:,end-1]
    v_new[:,end] = v[:,2]
    v_new[1,:] = v[end-1,:]
    v_new[end,:] = v[2,:]
end

function gray_scott()
    anim = Animation()

    N = 100
    dims = (N, N)

    FT = Float64
    args = Params()

    n_steps = 20000 # number of steps to take
    frame_interval = 200 # steps to take between making plots


    # u = cuNumeric.zeros(dims)
    # v = cuNumeric.zeros(dims)

    # u_new = cuNumeric.zeros(dims)
    # v_new = cuNumeric.zeros(dims)

    u = ones(FT, dims)
    v = zeros(FT, dims)

    u_new = zeros(FT, dims)
    v_new = zeros(FT, dims)

    u[1:15,1:15] = rand(FT, (15,15))
    v[1:15,1:15] = rand(FT, (15,15))

    for n in 1:n_steps
        step(u, v, u_new, v_new, args)
        # update u and v 
        # this doesn't copy, this switching references 
        u, u_new = u_new, u
        v, v_new = v_new, v

        if n%frame_interval == 0
            # plot
            heatmap(u, clims=(minimum(u), maximum(u)))
            frame(anim)
        end
    end
    gif(anim, "gray-scott.gif", fps=10)

end

gray_scott()

File: examples/gray-scott.py
# python equivalent of gray-scott.jl to test the GC problem

import cupynumeric as np


# import matplotlib.animation as animation
# from IPython.display import HTML
# import matplotlib.pyplot as plt

def greyScottSys(u, v, dx, dt, c_u, c_v, f, k):
    # u,v are arrays 
    # dx,dt are space and time steps
    # c_u, c_v, f, k are constant paramaters
    
     #create new u array
    u_new = np.zeros_like(u)
    v_new = np.zeros_like(v)
    
    #calculate F_u and F_v functions
    F_u = (-u[1:-1,1:-1]*(v[1:-1,1:-1]**2)) + f*(1-u[1:-1,1:-1])
    F_v = (u[1:-1,1:-1]*(v[1:-1,1:-1]**2)) - (f+k)*v[1:-1,1:-1]
    
    # 2-D Laplacian of f using array slicing, excluding boundaries
    # For an N x N array f, f_lap is the N-1 x N-1 array in the "middle"
    u_lap = (u[2:,1:-1] - 2*u[1:-1,1:-1] + u[:-2,1:-1]) / dx**2\
            + (u[1:-1,2:] - 2*u[1:-1,1:-1] + u[1:-1,:-2]) / dx**2
    v_lap = (v[2:,1:-1] - 2*v[1:-1,1:-1] + v[:-2,1:-1]) / dx**2\
            + (v[1:-1,2:] - 2*v[1:-1,1:-1] + v[1:-1,:-2]) / dx**2

    # Forward-Euler time step for all points except the boundaries
    u_new[1:-1,1:-1] = ((c_u * u_lap) + F_u)*dt + u[1:-1,1:-1]
    v_new[1:-1,1:-1] = ((c_v * v_lap) + F_v)*dt + v[1:-1,1:-1]

    # Apply periodic boundary conditions
    u_new[:,0] = u[:,-2]
    u_new[:,-1] = u[:,1]
    u_new[0,:] = u[-2,:]
    u_new[-1,:] = u[1,:]
    v_new[:,0] = v[:,-2]
    v_new[:,-1] = v[:,1]
    v_new[0,:] = v[-2,:]
    v_new[-1,:] = v[1,:]

    return u_new, v_new



# initial conditions and discretizaiton
dx = 1
dt = dx/5
u = np.ones((1000,1000))
v = np.zeros((1000,1000))
u[:150,:150] = np.random.rand(150,150)
v[:150,:150] = np.random.rand(150,150)


# fig = plt.figure()

c_u = 1
c_v = 0.3
f = 0.03
k = 0.06

# t_final = 1000

# ims = []
n_steps = 2000 # number of steps to take
frame_interval = 200 # steps to take between making plots

# build a list of images
for n in range(n_steps) :
    
    ## This may need to be changed.
    u,v = greyScottSys(u, v, dx, dt, c_u, c_v, f, k)
    
    # ## Store frames when n is a multiple of frame_interval
    # if n%frame_interval == 0:
    #     im = plt.imshow(u, vmin=0, vmax=1) # Show a plot of u.
    #     ims.append([im]) # append single image to the list of images

# anim = animation.ArtistAnimation(fig, ims, interval=100, repeat=False)
# HTML(anim.to_jshtml())
